src/interface/comm.rs

```diff
@@ -1,18 +1,281 @@
-// // Author: Nicholas Renner
+// // Authors: Nicholas Renner and Jonathan Singer
 // //
 // //
 
-// use std::collections::{HashMap, HashSet};
+use std::mem::size_of;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::fs::read_to_string;
+extern crate libc;
 
+static NET_DEV_FILENAME: &str = "net_devices";
 
-// static _BOUND_SOCKETS: HashMap = HashMap::new();
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+pub enum GenSockaddr {
+    V4(SockaddrV4),
+    V6(SockaddrV6)
+}
+impl GenSockaddr {
+    pub fn port(&self) -> u16 {
+        match self {
+            GenSockaddr::V4(v4addr) => v4addr.sin_port,
+            GenSockaddr::V6(v6addr) => v6addr.sin6_port
+        }
+    }
+    pub fn set_port(&mut self, port: u16) {
+        match self {
+            GenSockaddr::V4(v4addr) => v4addr.sin_port = port,
+            GenSockaddr::V6(v6addr) => v6addr.sin6_port = port
+        };
+    }
 
-// static OPEN_SOCKET_INFO: HashMap = HashMap::new();
+    pub fn addr(&self) -> GenIpaddr {
+        match self {
+            GenSockaddr::V4(v4addr) => GenIpaddr::V4(v4addr.sin_addr),
+            GenSockaddr::V6(v6addr) => GenIpaddr::V6(v6addr.sin6_addr)
+        }
+    }
+    pub fn set_addr(&mut self, ip: GenIpaddr){
+        match self {
+            GenSockaddr::V4(v4addr) => v4addr.sin_addr = if let GenIpaddr::V4(v4ip) = ip {v4ip} else {unreachable!()},
+            GenSockaddr::V6(v6addr) => v6addr.sin6_addr = if let GenIpaddr::V6(v6ip) = ip {v6ip} else {unreachable!()}
```

Comment:
I mean it may be somewhat overkill but this really should be an entirely unreachable code path that the user should not be able to trigger and that's usually where you want to panic.